tests/__serialize_015.phpt

```diff
@@ -0,0 +1,61 @@
+--TEST--
+__serialize() mechanism (015): Uninitialized properties from __sleep can be serialized and unserialized
+--SKIPIF--
+<?php
+if (PHP_VERSION_ID < 70400) { echo "skip __serialize/__unserialize not supported in php < 7.4 for compatibility with serialize()"; }
+if (PHP_VERSION_ID >= 80000) { echo "skip different error message format"; }
+?>
+--FILE--
+<?php
+error_reporting(E_ALL);
+set_error_handler(function ($errno, $message) {
+    echo $message . "\n";
+});
+class OSI {
+    public stdClass $o;
+    public string $s;
+    public ?int $i;
+    public float $f;
+    public function __sleep() {
+        return ['o', 's', 'i'];
+    }
+}
+// 00000002               -- header
+// 17 03 4d79436c617373   -- object of type "MyClass"
+//   14 03 000000           -- with 3 uninitialized properties
+$m = new OSI();
+var_dump($m);
+var_dump($s = serialize($m));
+try {
+    var_dump(unserialize($s));
+} catch (Error $e) {
+    // TODO: Double check if this is a deliberate design decision.
+    echo "unserialize: {$e->getMessage()}\n";
+}
+var_dump(bin2hex($s = igbinary_serialize($m)));
+try {
+    var_dump(igbinary_unserialize($s));
+} catch (Error $e) {
+    echo "igbinary_unserialize: {$e->getMessage()}\n";
+}
+--EXPECT--
+object(OSI)#2 (0) {
+  ["o"]=>
+  uninitialized(stdClass)
+  ["s"]=>
+  uninitialized(string)
+  ["i"]=>
+  uninitialized(?int)
+  ["f"]=>
+  uninitialized(float)
+}
+serialize(): "o" returned as member variable from __sleep() but does not exist
+serialize(): "s" returned as member variable from __sleep() but does not exist
+serialize(): "i" returned as member variable from __sleep() but does not exist
+string(44) "O:3:"OSI":3:{s:1:"o";N;s:1:"s";N;s:1:"i";N;}"
```

Comment:
I didn't see any mention of how uninitialized typed properties should behave for `__sleep` in the RFC or RFC/PR discussion. Right now, because igbinary's `igbinary_serialize()` is based on php's `serialize()` (which serializes nulls), it has the exact same TypeError on `igbinary_unserialize()`

Is the current behavior in 7.4 and 8.0-dev for __sleep on uninitialized properties by design, or is it an edge case that was overlooked? It seems surprising that data serializes successfully, but throws when it gets unserialized for the same class declaration.

- I checked https://wiki.php.net/rfc/typed_properties_v2 , https://externals.io/message/102333 , and https://externals.io/message/103148 , and didn't see any references to __sleep (just "I wouldn't lose sleep
if we wanted to keep the GA date where it is.")

I also didn't see any mention of `__sleep` in https://github.com/php/php-src/pull/3734 , or in the changes of the PR. (or in https://bugs.php.net/search.php?cmd=display&search_for=__sleep+AND+uninitialized&order_by=ts1&direction=ASC&limit=30&status=Open&reorder_by=ts2 )

cc @nikic @bwoeibi @krakjoe @dstogov (co-authors of `https://github.com/php/php-src/commit/e219ec144ef6682b71e135fd18654ee1bb4676b4`)